lightning/src/ln/channelmanager.rs

```diff
@@ -8818,6 +8818,22 @@ where
 		};
 		let invoice_request = builder.build_and_sign()?;
 
+		let total_liquidity: u64 = self.list_usable_channels().iter().map(|channel| channel.next_outbound_htlc_limit_msat).sum();
+		let total_amount_msats = match invoice_request.amount_msats() {
+			Some(amount_msats) => Some(amount_msats),
+			None => match offer.amount() {
+				Some(Amount::Bitcoin { amount_msats }) => Some(amount_msats),
+				_ => None,
+			},
+		};
+
+		if let Some(amount) = total_amount_msats {
+			if amount > total_liquidity {
+				log_error!(self.logger, "Insufficient liquidity for payment with payment id: {}", payment_id);
+				return Err(Bolt12SemanticError::InsufficientLiquidity);
+			}
+		}
```

Comment:
Let's add a test to `offers_tests.rs` to check this error. Looks like one of those tests is failing when hitting this case, so we'll likely need to modify that test setup.

Reply:
Hmmm... the test is checking `DefaultMessageRouter`'s  behavior, so it would be nice to keep that coverage. I think the only way to bypass the liquidity check is give the offer a fiat denomination using `OfferBuilder::amount`  (making it `pub(crate)`) instead of `OfferBuilder::amount_msats`.

Reply:
So I was helping @slanesuke debug why this isn't working, and it seems we currently don't support _paying_ fiat-denominated offers, which is unexpected -- we should only not support _creating_ them. `InvoiceRequestBuilder::build_and_sign` is failing because it eventually calls `OfferContents::check_amount_msats_for_quantity`, which is returning `UnsupportedCurrency`.

That function is called when parsing an `InvoiceRequest`, too, so we'd want to fail there, I think. Maybe we move the currency check to the parsing code and let the builder code trust the msats amount set by the user?

Reply:
> That function is called when parsing an `InvoiceRequest`, too, so we'd want to fail there, I think. Maybe we move the currency check to the parsing code and let the builder code trust the msats amount set by the user?

@slanesuke Let me know if you can take this on. The solution above should be right.